TRAMP improvements

vemv opened the pull request:
* Add a Dockerfile for TRAMP development purposes
  * Includes a README.
* Make TRAMP functionality work when using non-standard ports
  * e.g. when we connect from a SSH port other than `22`.
* Don't lose the CIDER session over TRAMP files
  * Makes `M-.` work recursively (instead of just once) when using TRAMP.
  * Fixes https://github.com/clojure-emacs/cider/issues/3250

bbatsov commented:
The changes look good overall. Do we also need to update/extend https://docs.cider.mx/cider/basics/up_and_running.html#working-with-remote-hosts ?

vemv commented:
Hey @mmontone @igrishaev ,

`No linked CIDER sessions` should be far more rare to encounter now (as of CIDER master, as released already into MELPA non- `Stable`)

* For vanilla CIDER connections, it generally shouldn't happen
* For TRAMP connections, we fixed the issue
* For Docker-backed connections, we also fixed the issue

It would be much helpful if you could try removing the workaround and checking if it works now.

If it didn't, you now can:

* Try jumping to definition,
* See it fail with `No linked CIDER sessions`
* Then, use the new command `M-x cider-debug-sesman-friendly-session-p`
* Copy its output (as visible in the `*messages*` buffer) and paste it here.

Thanks much - V
